src/Npgsql.NodaTime/DateHandler.cs

```diff
@@ -65,5 +71,29 @@ public override void Write(LocalDate value, NpgsqlWriteBuffer buf, NpgsqlParamet
             var totalDaysSinceEra = Period.Between(default(LocalDate), value, PeriodUnits.Days).Days;
             buf.WriteInt32(totalDaysSinceEra - 730119);
         }
+
+        protected internal override int ValidateObjectAndGetLength(object value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
```

Comment:
There are too many methods in type handlers which makes it hard to understand where is the actual entry point, so I perfectly understand your confusion here (:

What you should touch is internal methods having the `TAny` generic parameter: `Read<TAny>`, `ValidateAndGetLength<TAny>` and `WriteWithLengthInternal<TAny>`. These methods are responsible for strongly typed read and write operations and they are non-allocating. Internally they call corresponding methods of `INpgsqlTypeHandler<T>` if the class implements `INpgsqlTypeHandler<TAny>`.

Other three entry points are non-generic methods `ReadAsObject`, `ValidateObjectAndGetLength` and `WriteObjectWithLength`. Since they work with objects they allocate memory when the parameter or the default type is a value type because of boxing as you surely understand. Also the non-generic write is slower that the generic version because it calls a delegate pointing to `WriteWithLengthInternal<TAny>` in a generated `if (value is T) WriteWithLengthInternal<T>(/* parameters */);`.

Therefore, to support all scenarios all generic and non-generic methods should check the argument and decide what handler should be used, but this costs us an unmaintainable code (it really smells, take a look at #3092). Plus this interception decreases performance for NodaTime types which we don't want. So let's implement interfaces that the corresponding type handlers for BCL have and delegate handling in method implementations like you did in this handler. The entry point methods should be removed.

Reply:
I think this is simpler... The entry points that work with objects, as well as those that are generic, are designed to be implemented by type handlers which handle an open set of types in some way. For example, the JSON type handler can handle any .NET type, and the array type handler handles arbitrarily dimensioned arrays (`int[]`. `int[,]`, `int[,,]`...).

In the simpler case, type handlers with a closed set of supported types - like this one - should just implement INpgsqlTypeHandler (or INpgsqlSimpleTypeHandler) for each supported type; no need for anything more complicated. So just like you've implemented `INpgsqlSimpleTypeHandler<NpgsqlDate>`, the type handler can implement `INpgsqlSimpleTypeHandler<DateTime>`, and the methods coming from that should make everything work easily.

Reply:
You simplified what I tried to say, it looks much clear (:

Reply:
Oh great :)

I agree the type handler system right now is a bit... complicated (maybe even convoluted). It evolved this way over several years. I actually think there's an ok logic - the typical/regular type handler just implements the interface for the types it wants to support, and extends NpgsqlTypeHandler for the "default" type. But it's true there's too much indirection going on, making it both confusing and less performant.